### PR TITLE
[6203] Make default Preview edit mode on/off configurable

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/6203

### CODE REVIEW

File: ui/app/src/services/configuration.ts [hunk 1]
```diff
@@ -270,6 +270,9 @@ export function fetchSiteConfig(site: string, environment: string): Observable<S
         .filter(Boolean),
       upload: ((node) => (node ? deserialize(node).upload : {}))(dom.querySelector(':scope > upload')),
       locale: ((node) => (node ? deserialize(node).locale : {}))(dom.querySelector(':scope > locale')),
+      defaultEditMode: ((node) => (node ? deserialize(node).defaultEditMode : null))(
+        dom.querySelector(':scope > defaultEditMode')
+      ),
```
Comment (rart) [inline ui/app/src/services/configuration.ts]: This needs to go to ui.xml. Create a new block:
```xml
<widget 
  id="craftercms.components.Preview" 
  initialEditModeOn="true|false"
  initialHighlightMode="edit|move"
/>
```

File: ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx [hunk 2]
```diff
@@ -500,6 +500,8 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
     const localEditMode = getStoredEditModeChoice(user.username);
     if (nnou(localEditMode) && editMode !== localEditMode) {
       dispatch(setPreviewEditMode({ editMode: localEditMode }));
+    } else if (nou(localEditMode) && nnou(defaultEditMode) && editMode !== defaultEditMode) {
+      dispatch(setPreviewEditMode({ editMode: defaultEditMode }));
```
Comment (rart) [inline ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx]: I think the idea would be to have a `initPreviewConfig` similar to all other config sets and that would set things at the reducer level. The components never look at the _initial_ config value.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/state/reducers/preview.ts [hunk 3]
```diff
@@ -190,6 +191,28 @@ const fetchGuestModelsCompleteHandler = (state, { type, payload }) => {
...
 
 const reducer = createReducer<GlobalState['preview']>(initialState, {
+  [initPreviewConfig.type]: (state, { payload }) => {
+    const configDOM = fromString(payload.configXml);
+    const initialEditModeOn = configDOM
+      .querySelector('[id="craftercms.components.Preview"]')
+      ?.getAttribute('initialEditModeOn');
+    const initialHighlightMode = configDOM
+      .querySelector('[id="craftercms.components.Preview"]')
+      ?.getAttribute('initialHighlightMode');
```
Comment (rart) [inline ui/app/src/state/reducers/preview.ts]:
```suggestion
const previewConfigEl = configDOM
      .querySelector('[id="craftercms.components.Preview"]')
    const initialEditModeOn = previewConfigEl
      ?.getAttribute('initialEditModeOn');
    const initialHighlightMode = previewConfigEl
      ?.getAttribute('initialHighlightMode');
```

File: ui/app/src/state/reducers/preview.ts [hunk 4]
```diff
@@ -190,6 +191,28 @@ const fetchGuestModelsCompleteHandler = (state, { type, payload }) => {
...
+
+    const editMode = payload.storedEditMode ?? (nnou(initialEditModeOn) ? initialEditModeOn === 'true' : null);
```
Comment (rart) [inline ui/app/src/state/reducers/preview.ts]:
```suggestion
    state.editMode = payload.storedEditMode ?? (initialEditModeOn ? initialEditModeOn === 'true' : state.editMode);
```

File: ui/app/src/state/reducers/preview.ts [hunk 5]
```diff
@@ -190,6 +191,28 @@ const fetchGuestModelsCompleteHandler = (state, { type, payload }) => {
...
+    const highlightMode = payload.storedHighlightMode ?? initialHighlightMode;
```
Comment (rart) [inline ui/app/src/state/reducers/preview.ts]:
```suggestion
    state.highlightMode = payload.storedHighlightMode ?? (['all', 'move'].includes(initialHighlightMode) ? initialHighlightMode : state.highlightMode);
```

File: ui/app/src/state/reducers/preview.ts [hunk 6]
```diff
@@ -190,6 +191,28 @@ const fetchGuestModelsCompleteHandler = (state, { type, payload }) => {
...
+
+    if (nnou(editMode)) {
+      state.editMode = editMode;
+    }
+    if (nnou(highlightMode)) {
+      state.highlightMode = highlightMode;
+    }
+    if (nnou(payload.storedPaddingMode)) {
+      state.editModePadding = payload.storedPaddingMode;
+    }
```
Comment (rart) [inline ui/app/src/state/reducers/preview.ts]:
```suggestion
    state.editModePadding = payload.storedPaddingMode ?? state.editModePadding;
```

File: ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx [hunk 7]
```diff
@@ -494,26 +502,9 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
...
-    if (nnou(localHighlightMode) && highlightMode !== localHighlightMode) {
-      dispatch(setHighlightMode({ highlightMode: localHighlightMode }));
-    }
-
-    const localPaddingMode = getStoredEditModePadding(user.username);
-    if (nnou(localPaddingMode) && editModePadding !== localPaddingMode) {
-      dispatch(setEditModePadding({ editModePadding: localPaddingMode }));
-    }
-
     startCommunicationDetectionTimeout(guestDetectionTimeoutRef, setGuestDetectionSnackbarOpen);
```
Comment (rart) [inline ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx]: Let's add `xbDetectionTimeoutMs` too as a possible config. Make 0 not check at all.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/state/reducers/preview.ts (hunk 5 above)
Comment (rart) [reply ui/app/src/state/reducers/preview.ts]: Please confirm the correct value list for `all`, `move`

Comment (jvega190) [reply ui/app/src/state/reducers/preview.ts]: `all` and `move` are correct

**Comment by rart** [comment]
Conflicts.

File: ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx [hunk 8]
```diff
@@ -450,13 +459,26 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
...
+      const storedEditMode = getStoredEditModeChoice(user.username, uuid);
+      const storedHighlightMode = getStoredHighlightModeChoice(user.username, uuid);
+      const storedPaddingMode = getStoredEditModePadding(user.username);
+      dispatch(initPreviewConfig({ configXml: uiConfig.xml, storedEditMode, storedHighlightMode, storedPaddingMode }));
+    }
+  }, [uiConfig.xml, user.username, uuid, dispatch]);
+
+  useEffect(() => {
+    if (!socketConnected && authActive && xbDetectionTimeoutMs) {
+      startCommunicationDetectionTimeout(
```
Comment (rart) [inline ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx]: The socket connection should stay as it was. Only XB detection should take this change.

File: ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx [hunk 9]
```diff
@@ -494,26 +516,14 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
...
-      dispatch(setEditModePadding({ editModePadding: localPaddingMode }));
+  useEffect(() => {
+    if (xbDetectionTimeoutMs) {
+      startCommunicationDetectionTimeout(guestDetectionTimeoutRef, setGuestDetectionSnackbarOpen, xbDetectionTimeoutMs);
     }
+  }, [xbDetectionTimeoutMs]);
 
-    startCommunicationDetectionTimeout(guestDetectionTimeoutRef, setGuestDetectionSnackbarOpen);
-
+  // Guest detection, document domain restoring and guest key up/down notifications.
```
Comment (rart) [inline ui/app/src/components/PreviewConcierge/PreviewConcierge.tsx]:
```suggestion
  // Document domain restoring.
```

File: ui/app/src/state/reducers/preview.ts [hunk 10]
```diff
@@ -238,6 +241,23 @@ const fetchGuestModelsCompleteHandler = (state, { type, payload }) => {
...
+    const initialHighlightMode = previewConfigEl?.getAttribute('initialHighlightMode');
+    const xbDetectionTimeoutMs = parseInt(previewConfigEl?.getAttribute('xbDetectionTimeoutMs'));
+
+    state.editMode = payload.storedEditMode ?? (initialEditModeOn ? initialEditModeOn === 'true' : state.editMode);
+    state.highlightMode =
+      payload.storedHighlightMode ??
+      (['all', 'move'].includes(initialHighlightMode) ? initialHighlightMode : state.highlightMode);
+    state.editModePadding = payload.storedPaddingMode ?? state.editModePadding;
+    state.xbDetectionTimeoutMs =
+      Boolean(xbDetectionTimeoutMs) && xbDetectionTimeoutMs > 0
```
Comment (rart) [inline ui/app/src/state/reducers/preview.ts]: `0` should be a valid value which means to not execute the check.
```suggestion
      nnou(xbDetectionTimeoutMs)
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/state/reducers/preview.ts [hunk 11]
```diff
@@ -238,6 +241,23 @@ const fetchGuestModelsCompleteHandler = (state, { type, payload }) => {
...
+    const initialHighlightMode = previewConfigEl?.getAttribute('initialHighlightMode');
+    const xbDetectionTimeoutMs = parseInt(previewConfigEl?.getAttribute('xbDetectionTimeoutMs'));
+
+    state.editMode = payload.storedEditMode ?? (initialEditModeOn ? initialEditModeOn === 'true' : state.editMode);
+    state.highlightMode =
+      payload.storedHighlightMode ??
+      (['all', 'move'].includes(initialHighlightMode) ? initialHighlightMode : state.highlightMode);
+    state.editModePadding = payload.storedPaddingMode ?? state.editModePadding;
+    state.xbDetectionTimeoutMs =
+      !isNaN(xbDetectionTimeoutMs) && xbDetectionTimeoutMs > 0
```
Comment (rart) [inline ui/app/src/state/reducers/preview.ts]:
```suggestion
      !isNaN(xbDetectionTimeoutMs) && xbDetectionTimeoutMs >= 0
```

**Review by rart — CHANGES_REQUESTED** [review]

**Review by rart — APPROVED** [review]

**Pull request merged**